allennlp/nn/util.py

```diff
@@ -1264,6 +1264,89 @@ def batched_index_select(
     return selected_targets
 
 
+def batched_index_fill(
+    target: torch.Tensor, indices: torch.LongTensor, mask: torch.Tensor, fill_value: int = 1
+) -> torch.Tensor:
+    """
+    The given `indices` in `target` will be will be filled with `fill_value` given a `mask`.
+
+
+    # Parameters
+
+    target : `torch.Tensor`, required.
+        A 2 dimensional tensor of shape (batch_size, sequence_length).
+        This is the tensor to be filled.
+    indices : `torch.LongTensor`, required
+        A 2 dimensional tensor of shape (batch_size, num_indices),
+        These are the indices that will be filled in the original tensor.
+    mask : `torch.Tensor`, required.
+        A 2 dimensional tensor of shape (batch_size, num_indices), mask.sum() == `nonzero_indices`.
+    fill_value : `int`, optional (default = `1`)
+        The value we fill the tensor with.
+
+    # Returns
+
+    filled_target : `torch.Tensor`
+        A tensor with shape (batch_size, sequence_length) where 'indices' are filled with `fill_value`
+    """
+    prev_shape = target.size()
+    # Shape: (batch_size * num_indices)
+    flattened_indices = flatten_and_batch_shift_indices(indices * mask, target.size(1))
+    # Shape: (batch_size * num_indices, 1)
+    mask = mask.bool().view(-1)
+    # Shape: (batch_size * sequence_length, 1)
+    flattened_target = target.view(-1, 1)
+    # Shape: (nonzero_indices, 1)
+    unmasked_indices = flattened_indices[mask].unsqueeze(-1)
+
+    flattened_target.scatter_(0, unmasked_indices, fill_value)
+
+    filled_target = flattened_target.reshape(prev_shape)
+
+    return filled_target
+
+
+def batched_index_scatter(
+    target: torch.Tensor, indices: torch.LongTensor, mask: torch.Tensor, replace: torch.Tensor,
+) -> torch.Tensor:
+    """
+    The given `indices` in `target` will be will be scattered with corresponding index
+    from the `replace` tensor given a `mask`.
```

Comment:
In a way, this is `target[indices] = replace[indices]`, but with a mask? Is an operation like that really called "scatter"? It sounds more like `batched_index_replace` to me.

![images](https://user-images.githubusercontent.com/920638/91286260-12a74a80-e743-11ea-8e81-7dd5701e04bd.jpeg)

I have checked the pytorch docs, and it is indeed called "scatter": https://pytorch.org/docs/stable/tensors.html#torch.Tensor.masked_scatter_

I'm surprised this function doesn't boil down to `masked_scatter()` or `index_copy()`.

Reply:
Yeah, they do, but they don't deal with both masks and indices, I guess better names would be masked_index_fill and masked_index_replace

Reply:
I'm happy with whatever names you think make sense :-)